### PR TITLE
php APCu Integration.

### PR DESCRIPTION
**Pull request by withgod**
### What does this PR do?

Add APCu (a major cache in php) integration.

### Motivation

We wanted to add them in common instead of preparing them separately for several teams.

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [ ] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

I communicated with support before the PR, but it was going in circles, so I decided it would be faster to have the code reviewed.

* https://help.datadoghq.com/hc/en-us/requests/417049

### CODE REVIEW

File: php_apcu/datadog_checks/php_apcu/check.py
Comment (ChristineTChen) [inline php_apcu/datadog_checks/php_apcu/check.py:15]: The Agent base package support a request wrapper that would support additional options when making calls like timeout, proxies, tls verification, etc.

It's quite easy to set up, you just have to replace `request.get` with `self.http.get`. 

You can read more about it here: https://datadoghq.dev/integrations-core/base/http/

File: php_apcu/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline php_apcu/assets/configuration/spec.yaml:26]:
```suggestion
        example: http://localhost/apcu-status
     - template: instances/http
     - template: instances/default
```
Missing default template for instance level.

If you implement the request wrapper as mentioned before, you can take advantage of the `http` config template.

File: php_apcu/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline php_apcu/assets/configuration/spec.yaml:7]:
```suggestion
    - template: init_config/http
    - template: init_config/default
```
same here

File: php_apcu/manifest.json
Comment (ChristineTChen) [inline php_apcu/manifest.json:7]:
```suggestion
  "metric_to_check": "php_apcu.cache.mem_size",
```

Including a metric to check will auto-install this tile when metrics are detected

File: php_apcu/README.md
Comment (ChristineTChen) [inline php_apcu/README.md:74]: Link to the php_apcu site

**Review by ChristineTChen — CHANGES_REQUESTED** [review]
The CI is missing your integration, you can fix by running `ddev validate ci --fix`

Additionally, can you add yourself as the integration owner in the CODEOWNERS file? https://github.com/DataDog/integrations-extras/blob/master/.github/CODEOWNERS

Thanks!

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md]: fixed 28cdd6d7124fbadf91bb9b3e9f8ab47a1d1093df

File: php_apcu/manifest.json
Comment (withgod) [reply php_apcu/manifest.json]: fixed c4935f725130efef0b2ec0ec5b3c0a44a7ac860d

File: php_apcu/datadog_checks/php_apcu/check.py
Comment (withgod) [reply php_apcu/datadog_checks/php_apcu/check.py]: fixed 7277ab326a70898fad0201f53ef010d6298e3d25

File: php_apcu/assets/configuration/spec.yaml
Comment (withgod) [reply php_apcu/assets/configuration/spec.yaml:27]: same fixed 7277ab326a70898fad0201f53ef010d6298e3d25

File: php_apcu/assets/configuration/spec.yaml
Comment (withgod) [reply php_apcu/assets/configuration/spec.yaml:8]: same fixed 7277ab326a70898fad0201f53ef010d6298e3d25

**Comment by withgod** [comment]
>The CI is missing your integration, you can fix by running ddev validate ci --fix

I was tried `ddev validate ci --fix`, but found other integration diff.

https://gist.github.com/withgod/6cbd9a458f7c64a9933cd1e1a487fc40#file-ddev-validate-ci-fix-log-L32-L56

wondering whether to commit.
It's an automatically generated one, so if the format is broken, I think it's okay to overwrite it...

**Comment by ChristineTChen** [comment]
> > The CI is missing your integration, you can fix by running ddev validate ci --fix
> 
> I was tried `ddev validate ci --fix`, but found other integration diff.
> 
> https://gist.github.com/withgod/6cbd9a458f7c64a9933cd1e1a487fc40#file-ddev-validate-ci-fix-log-L32-L56
> 
> wondering whether to commit.
> It's an automatically generated one, so if the format is broken, I think it's okay to overwrite it...

Looks to be just an indentation fix, maybe that was added manually. I think its okay to commit. Thanks for double checking!

**Comment by withgod** [comment]
i was fixed review and checks all green.
please check @ChristineTChen.
thanks.

File: php_apcu/README.md
Comment (kayayarai) [inline php_apcu/README.md:36]:
```suggestion
To start collecting APCu metrics, you must configure the URL for the APCu Datadog handler in the `conf.yaml` file:
```

File: php_apcu/assets/monitors/php-apcu_expunges.json
Comment (kayayarai) [inline php_apcu/assets/monitors/php-apcu_expunges.json:5]:
```suggestion
    "message": "APCu Cache Full has been detected. All cache purged.",
```

File: php_apcu/README.md
Comment (kayayarai) [inline php_apcu/README.md:37]:
```
instances
  - url: http://localhost/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/php_apcu/assets/exporter/apcu-dd-handler.php
```

Or, create an alias to the handler file:

File: php_apcu/README.md
Comment (kayayarai) [inline php_apcu/README.md:45]: and add configure it in the `conf.yaml` file:
```
instances
  - url: http://localhost/apcu-status
```

File: php_apcu/assets/configuration/spec.yaml
Comment (kayayarai) [inline php_apcu/assets/configuration/spec.yaml:17]: update with the edits I suggested to the README file

File: php_apcu/datadog_checks/php_apcu/data/conf.yaml.example
Comment (kayayarai) [inline php_apcu/datadog_checks/php_apcu/data/conf.yaml.example:52]: Update with the edits I suggested to the README file.

**Review by kayayarai — CHANGES_REQUESTED** [review]
Made some significant suggestions to the readme instructions. I hope I got the technical details correct.

File: php_apcu/manifest.json
Comment (withgod) [reply php_apcu/manifest.json]: fixed 96b2f99f2ed366ca518abc9c519f103827430049

File: php_apcu/manifest.json
Comment (withgod) [reply php_apcu/manifest.json]: fixed 96b2f99f2ed366ca518abc9c519f103827430049

File: php_apcu/manifest.json
Comment (withgod) [reply php_apcu/manifest.json]: fixed 96b2f99f2ed366ca518abc9c519f103827430049

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md]: fixed 96b2f99f2ed366ca518abc9c519f103827430049

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md:37]: tell me more about what to fix and how to fix it.

url: example you provided is the URL used if you set the DocumentRoot to /.

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md:45]: already going to provide that URL.
https://github.com/withgod/integrations-extras/blob/feature/apcu/php_apcu/assets/configuration/spec.yaml#L11-L27
https://github.com/withgod/integrations-extras/blob/feature/apcu/php_apcu/datadog_checks/php_apcu/data/conf.yaml.example#L61

File: php_apcu/datadog_checks/php_apcu/check.py
Comment (ChristineTChen) [inline php_apcu/datadog_checks/php_apcu/check.py:8]:
```suggestion
            raise ConfigurationError('The `url` must be specified')
```

File: php_apcu/datadog_checks/php_apcu/check.py
Comment (ChristineTChen) [inline php_apcu/datadog_checks/php_apcu/check.py:27]:
```suggestion
                    self.gauge(metric, value)
                else:
                    self.log.debug("Unexpected response: %s", values)
```
Let's add a debug line for what gets skipped, easier to debug for future

File: php_apcu/tests/test_php_apcu.py
Comment (ChristineTChen) [inline php_apcu/tests/test_php_apcu.py:24]: Can we assert some metrics?

File: php_apcu/tests/test_php_apcu.py
Comment (ChristineTChen) [inline php_apcu/tests/test_php_apcu.py:28]: Let's put this in another test

File: php_apcu/tox.ini
Comment (ChristineTChen) [inline php_apcu/tox.ini:9]:
```suggestion
description =
    py38: e2e ready
ensure_default_envdir = true
```
This allows you to spin up the docker environment via `ddev env start <integration name> py38`

**Review by ChristineTChen — CHANGES_REQUESTED** [review]
Thanks for addressing my review, @withgod

I just left a couple small comments. This integration looks good from my end. I'll merge after docs is ✅  cc @kayayarai

File: php_apcu/tests/test_php_apcu.py
Comment (withgod) [reply php_apcu/tests/test_php_apcu.py:30]: I looked at other tests and implemented it.
but APCu is after start-up it is almost no data.
sample output https://gist.github.com/withgod/58fca0210443b52cbec45e36a4c19ad1

File: php_apcu/tox.ini
Comment (withgod) [reply php_apcu/tox.ini:11]: hi @ChristineTChen 
azure pipeline failed e2e test.
I think failed start at f75f0580a26434bfe8d48f814977158419404f31.

I'm not a frequent python(tox) user, so I don't know how to fix this.

File: php_apcu/tests/test_php_apcu.py
Comment (ChristineTChen) [reply php_apcu/tests/test_php_apcu.py:30]: That's alright, I think having some assertion is good. But if the data will be inconsistent, we can consider adding a sample test fixture to mock the process of metric. We do something similar in our core integrations, like so:

Here we save the output to `test/fixtures/` directory: https://github.com/DataDog/integrations-core/blob/master/cilium/tests/fixtures/agent_metrics.txt

We create a pytest fixture that mocks the response: https://github.com/DataDog/integrations-core/blob/master/cilium/tests/conftest.py#L79-L90

And in a test, we pass the fixture in the parameters and it'll mock the `request.get` call with the response and assert metric collection: https://github.com/DataDog/integrations-core/blob/master/cilium/tests/test_cilium.py#L9-L15

File: php_apcu/README.md
Comment (ChristineTChen) [reply php_apcu/README.md:35]: I believe this is addressed in the next section. This looks like the setup for APCu specifically, not the integration.

File: php_apcu/README.md
Comment (ChristineTChen) [inline php_apcu/README.md:35]:
```suggestion

APCu does not expose metrics by default. This integration includes a metric exporter, located here:

    ```
    /opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/php_apcu/assets/exporter/apcu-dd-handler.php
    ```

It will work if you refer to it as a php file, but you may want to configure it as follows
```

Let's reword this for clarity

File: php_apcu/README.md
Comment (ChristineTChen) [inline php_apcu/README.md:43]: Where is this part meant to be configured, is there a file or is this a command to run? In this path `/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/php_apcu/assets/exporter/apcu-dd-handler.php`?

If so let's reword this line to be more specific:
```
It will work if you refer to it as a php file, but you may want to configure it as follows
```

File: php_apcu/README.md
Comment (ChristineTChen) [inline php_apcu/README.md:48]:
```suggestion
    ```
    instances
      - url: http://localhost/apcu-status
    ```
```

Let's document an example conf.yaml in this section so users see the entire setup process.

**Review by ChristineTChen — COMMENTED** [review]
Clarified some documentation steps in README

File: php_apcu/tests/test_php_apcu.py
Comment (withgod) [reply php_apcu/tests/test_php_apcu.py:30]: I think it is a good mechanism to be able to use fixture.
However, in this case, I don't think we need to use fixture because we can get it from the URL via docker.

If you're using more complex values, you should use fixture to pass complex values, but in this case, it's just a simple way to send what you get.

I thought you might be interested in the fact that I compared it to 0, so I annotated it.
I'll refer to this when I collect more complex metrics in the future. 😄

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md:43]: The user needs to write it in the configuration file of apache.
To be precise, it needs to be written in the configuration file of the web server that the user is using.
However, since php does not depend on the web server, I am using the well-known Apache example.

The mod_status of apache also needs to be set, but it was just one line.
https://github.com/DataDog/integrations-core/tree/master/apache#installation
````
Install mod_status on your Apache servers and enable ExtendedStatus.
```

https://github.com/DataDog/integrations-core/tree/master/nginx#prepare-nginx
I wrote this in the README based on the nginx example.

>It will work if you refer to it as a php file, but you may want to configure it as follows
This file must be run through a web server, so writing it that way may cause confusion.

File: php_apcu/README.md
Comment (withgod) [reply php_apcu/README.md]: thanks 😄

File: php_apcu/README.md
Comment (kayayarai) [reply php_apcu/README.md:44]: Oooooh, okay. Thank you for this clarification. I am going to re-edit.

File: php_apcu/README.md
Comment (kayayarai) [inline php_apcu/README.md:30]:
```suggestion
APCu does not expose metrics by default so this integration includes a metric exporter, located here:
```

File: php_apcu/README.md
Comment (kayayarai) [inline php_apcu/README.md:36]:
```suggestion
When you configure your Agent (the `instances` setting, described next), you can refer to the exporter directly. by this file name, or you can configure an alias for it on your web server. For example, if you're using Apache, the alias in the web server configuration file would look like this:
```

File: php_apcu/README.md
Comment (kayayarai) [reply php_apcu/README.md:49]: I agree, add this example here.

File: php_apcu/assets/configuration/spec.yaml
Comment (kayayarai) [inline php_apcu/assets/configuration/spec.yaml:17]:
```suggestion
        Provide a URL to the metric exporter, either by pointing to the exporter file itself: 
        `DD_ROOT/embedded/lib/python3.8/site-packages/datadog_checks/php_apcu/assets/exporter/apcu-dd-handler.php` 
        or by configuring an alias in your web server configuration (Apache example shown):
```

**Review by kayayarai — CHANGES_REQUESTED** [review]

**Review by kayayarai — APPROVED** [review]
Documentation looks fantastic! Thank you for this.

**Review by ChristineTChen — APPROVED** [review]

**Pull request merged**
